src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,690 @@
+        /// and replace its m_WriteList arraylist field.
+        /// </summary>
+        private sealed class ConnectionArrayList : ArrayListWrapper
+        {
+            public ConnectionArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                if (s_connectionType.IsInstanceOfType(value))
+                {
+                    // Replace the HttpWebRequest arraylist inside this Connection object,
+                    // which allows us to intercept each new HttpWebRequest object added under
+                    // this Connection.
+                    ArrayList originalArrayList = s_writeListField.GetValue(value) as ArrayList;
+                    HttpWebRequestArrayList newArrayList = new HttpWebRequestArrayList(originalArrayList ?? new ArrayList());
+
+                    s_writeListField.SetValue(value, newArrayList);
+                }
+
+                return base.Add(value);
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for Connection.m_WriteList. The goal here is to
+        /// intercept all new HttpWebRequest objects being added to Connection.m_WriteList
+        /// and notify the listener about the HttpWebRequest that's about to send a request.
+        /// It also intercepts all HttpWebRequest objects that are about to get removed from
+        /// Connection.m_WriteList as they have completed the request.
+        /// </summary>
+        private sealed class HttpWebRequestArrayList : ArrayListWrapper
+        {
+            public HttpWebRequestArrayList(ArrayList list) : base(list)
+            {
+            }
+
+            public override int Add(object value)
+            {
+                HttpWebRequest request = value as HttpWebRequest;
+                if (request != null)
+                {
+                    s_instance.RaiseRequestEvent(request);
+                }
+
+                return base.Add(value);
+            }
+
+            public override void RemoveAt(int index)
+            {
+                HttpWebRequest request = base[index] as HttpWebRequest;
+                if (request != null)
+                {
+                    HttpWebResponse response = s_httpResponseAccessor(request);
+                    if (response != null)
+                    {
+                        s_instance.RaiseResponseEvent(request, response);
+                    }
+                }
+
+                base.RemoveAt(index);
+            }
+        }
+
+        #endregion
+
+        #region private methods
+
+        /// <summary>
+        /// Private constructor. This class implements a singleton pattern and only this class is allowed to create an instance.
+        /// </summary>
+        private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+        {
+        }
+
+        private void RaiseRequestEvent(HttpWebRequest request)
```

Comment:
I think option 1 is fine. We need to make a change to remove the temporary diagnosticlistener in there anyways, we can bundle this with that change. People are usually comfortable if you give good explanations and it's code deletion and not code addition.
Regarding Vance's comment about it being overweighted on Application Insights, I don't think that is true. Outside of correlation ID support, which is the motivation we have, the feature is merely notifying you when HttpWebRequest is about to send and receive. I think it can stand on its own.
Liudmila can look at how to add activity support. Nevertheless, I think we should get this change in first, and then make more changes on top of it.